core/invoker/src/main/scala/whisk/core/containerpool/kubernetes/KubernetesContainer.scala

```diff
@@ -0,0 +1,145 @@
+   */
+  def create(transid: TransactionId,
+             image: String,
+             userProvidedImage: Boolean = false,
+             environment: Map[String, String] = Map(),
+             labels: Map[String, String] = Map(),
+             name: Option[String] = None)(implicit kubernetes: KubernetesApi,
+                                          ec: ExecutionContext,
+                                          log: Logging): Future[KubernetesContainer] = {
+    implicit val tid = transid
+
+    val podName = name.getOrElse("").replace("_", "-").replaceAll("[()]", "").toLowerCase()
+    for {
+      id <- kubernetes.run(image, podName, environment, labels).recoverWith {
+        case _ => Future.failed(WhiskContainerStartupError(s"Failed to run container with image '${image}'."))
+      }
+      ip <- kubernetes.inspectIPAddress(id).recoverWith {
+        // remove the container immediately if inspect failed as
+        // we cannot recover that case automatically
+        case _ =>
+          kubernetes.rm(id)
+          Future.failed(WhiskContainerStartupError(s"Failed to obtain IP address of container '${id.asString}'."))
+      }
+    } yield new KubernetesContainer(id, ip)
+  }
+}
+
+/**
+ * Represents a container as run by kubernetes.
+ *
+ * This class contains OpenWhisk specific behavior and as such does not necessarily
+ * use kubernetes commands to achieve the effects needed.
+ *
+ * @constructor
+ * @param id the id of the container
+ * @param addr the ip & port of the container
+ */
+class KubernetesContainer(protected val id: ContainerId, protected val addr: ContainerAddress)(
+  implicit kubernetes: KubernetesApi,
+  protected val ec: ExecutionContext,
+  protected val logging: Logging)
+    extends Container {
+
+  /** The last read timestamp in the log file */
+  private val lastTimestamp = new AtomicReference[Option[String]](None)
+
+  // no-op under Kubernetes
+  def suspend()(implicit transid: TransactionId): Future[Unit] = Future.successful({})
+
+  // no-op under Kubernetes
+  def resume()(implicit transid: TransactionId): Future[Unit] = Future.successful({})
+
+  override def destroy()(implicit transid: TransactionId): Future[Unit] = {
+    super.destroy()
+    kubernetes.rm(id)
+  }
+
+  def logs(limit: ByteSize, waitForSentinel: Boolean)(implicit transid: TransactionId): Source[ByteString, Any] = {
+
+    val activationMarkerCheck: ByteString ⇒ Boolean = { bs ⇒
+      if (bs.containsSlice(DockerContainer.ActivationSentinel)) {
+        lastTimestamp.set(Some(bs.utf8String.parseJson.convertTo[LogLine].time))
+        true
+      } else {
+        false
+      }
+    }
+
+    kubernetes
+      .logs(id, lastTimestamp.get()) // todo - same sentinel check behavior as DockerContainer should be implemented?
+      .via(Framing.delimiter(delimiter, limit.toBytes.toInt))
+      .via(new CompleteAfterOccurrences(activationMarkerCheck, 2, waitForSentinel))
+      .recover {
+        case _: StreamLimitReachedException =>
```

Comment:
Is this Exception relevant? You dropped the limitWeighted stage. Was that deliberate? You'll read/store unbounded data.